### PR TITLE
Add support for `row_alias` syntax added in MySQL 8.0.19.

### PR DESCRIPTION
**Pull request by arthurschreiber**
## Description

This pull request adds support for the `row_alias` syntax added in MySQL 8.0.19 for `INSERT INTO` statements.

## Related Issue(s)

https://github.com/vitessio/vitess/issues/15509

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/sqlparser/sql.y [hunk 1]
```diff
@@ -84,6 +84,7 @@ func markBindVariable(yylex yyLexer, bvar string) {
   ctes          []*CommonTableExpr
   order         *Order
   limit         *Limit
+  rowAlias      RowAlias
```
Comment (arthurschreiber) [inline go/vt/sqlparser/sql.y:87]: Is there some convention around when to use a pointer vs a value here?

Comment (dbussink) [reply go/vt/sqlparser/sql.y:87]: Generally if it's optional, a pointer is usually better to indicate the absence / presence of the alias here.

File: go/vt/sqlparser/ast_format.go [hunk 2]
```diff
@@ -2010,6 +2010,18 @@ func (node OnDup) Format(buf *TrackedBuffer) {
 	buf.astPrintf(node, " on duplicate key update %v", UpdateExprs(node))
 }
 
+func (node RowAlias) Format(buf *TrackedBuffer) {
+	if node.TableName.IsEmpty() {
```
[the remark this answers is not in the document]
Comment (arthurschreiber) [reply go/vt/sqlparser/ast_format.go]: Done.

File: go/vt/sqlparser/sql.y (hunk 1 above)
Comment (arthurschreiber) [reply go/vt/sqlparser/sql.y]: Done.

**Comment by arthurschreiber** [comment]
Unit tests are failing due to some whitespace changes I accidentally committed. Will fix that later.

**Comment by arthurschreiber** [comment]
Looks like this increases the number of shift/reduce conflicts from 3 to 7? 🤔 I'm not quite sure why, if anyone could give me a few pointers on how to fix that, I'd be happy to take a stab at it.

**Comment by dbussink** [comment]
> Looks like this increases the number of shift/reduce conflicts from 3 to 7? 🤔 I'm not quite sure why, if anyone could give me a few pointers on how to fix that, I'd be happy to take a stab at it.

@GuptaManan100 is probably the subject matter expert on this topic.

**Comment by arthurschreiber** [comment]
Okay, I think I understand where the conflicts are coming from.

The `insert_data` parsing rule used by `insert_statement` can either be a list of values, or can be a `SELECT` from a different table. if it's a `SELECT`, it can contain `select_expression` which contains `as_ci_opt`, which conflicts with the `AS` used by the row alias syntax.

The good thing here is that the row alias syntax can only be used with `VALUES` or `SET`. This seems fixable, but will require some refactoring of the parser.

I also noticed that the way `insert_or_replace` is used can lead to invalid syntax being accepted by the Vitess parser. For example, `REPLACE` can not be used together with `ON DUPLICATE KEY UPDATE` - but the parser allows this.

**Comment by GuptaManan100** [comment]
I'm looking into it

**Review by GuptaManan100 — APPROVED** [review]

**Review by dbussink — APPROVED** [review]

**Review by frouioui — APPROVED** [review]

**Pull request merged**